File: src/geonhyeong/week2/문제집 1/1-2.py
#
#  1-2w.swift
#  Question Link:
#  Primary idea:   dictionary
#
#  Time Complexity : O()
#  Space Complexity : O()
#  Runtime:  ms
#  Memory Usage:  MB
#

from collections import defaultdict

def solution(want, number, discount):
    answer = 0

    for i in range(0, len(discount) - 9):
        dic = defaultdict(int)
        for food in discount[i:i+10]:
            dic[food] += 1

        flag = True
        for (j, w) in enumerate(want):
            if (not w in dic) or (dic[w] < number[j]):
                flag = False
                break
        
        if flag:
            answer += 1
                
    return answer

print(solution(["banana", "apple", "rice", "pork", "pot"], [3, 2, 2, 2, 1], ["chicken", "apple", "apple", "banana", "rice", "apple", "pork", "banana", "pork", "rice", "pot", "banana", "apple", "banana"])) # 3
print(solution(["apple"], [10], ["banana", "banana", "banana", "banana", "banana", "banana", "banana", "banana", "banana", "banana"])) # 0

File: src/taeeun/week2/1-2.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>

using namespace std;

int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;
    int days=discount.size()-10;
    map <string, int> products;
    for (int i=0; i<10; i++) {
        products[discount[i]]++;
    }
    for (int i=0; i<=days; i++) {
        bool flag=true;
        for (int j=0; j<want.size(); j++) {
            string want_product=want[j];
            int want_num=number[j];
            if (products[want_product] < want_num) {
                flag=false;
                break;
            }
        }
        if (flag) {
            answer++;
        }
        if (10+i<discount.size()) {
            products[discount[i]]--;
            products[discount[10+i]]++;
        }
    }

    return answer;
}
